### PR TITLE
feat(gatsby): forward jobs to main process from workers

### PR DESCRIPTION
**Pull request by pieh**
## Description

Building on messaging API to be added with https://github.com/gatsbyjs/gatsby/pull/32159 (against that PR branch). This route job creation from worker to main process so that it can handle completed or in-progress jobs, handle dupes etc. It builds on already existing message object shapes that are used for external jobs processing.

## Related Issues

[ch32062]

### CODE REVIEW

File: packages/gatsby/src/utils/jobs/manager.ts [hunk 1]
```diff
@@ -7,62 +7,23 @@ import _ from "lodash"
...
+import {
+  JobInput,
+  InternalJob,
+  MESSAGE_TYPES,
+  IJobCreatedMessage,
+  IJobCompletedMessage,
+  IJobFailed,
+  IJobNotWhitelisted,
+  WorkerError,
+} from "./types"
```
Comment (pieh) [inline packages/gatsby/src/utils/jobs/manager.ts:19]: Shared types moved do `./types.ts`. Also moved `jobs-manager.ts` to `jobs/manager.ts` to accomudate extra `jobs` modules

File: packages/gatsby/src/utils/jobs/manager.ts [hunk 2]
```diff
@@ -132,15 +93,19 @@ async function runLocalWorker<T>(
...
 function listenForJobMessages(): void {
   process.on(`message`, msg => {
-    if (
-      msg &&
-      msg.type &&
-      msg.payload &&
-      msg.payload.id &&
-      externalJobsMap.has(msg.payload.id)
-    ) {
+    if (isJobsIPCMessage(msg)) {
```
Comment (pieh) [inline packages/gatsby/src/utils/jobs/manager.ts:108]: jobs ipc messages are now typed now (this is code for external job runner), no change in behaviour

File: packages/gatsby/src/utils/jobs/worker-messaging.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    gatsbyWorkerMessenger.sendMessage(msg)
+
+    // holds on to promise
+    deferredWorkerPromises.set(job.id, deferredWorkerPromise)
+
+    return deferredWorkerPromise.promise
+  }
+
+  return undefined
+}
```
Comment (pieh) [inline packages/gatsby/src/utils/jobs/worker-messaging.ts:115]: Worker <-> Parent job related messages use same message types as Parent<->External job runner

File: packages/gatsby/src/utils/worker/__tests__/jobs.ts [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    void,
+    [msg: MessagesFromParent, workerId: number]
+  >
+  let receivedMessageSpy: jest.Mock<
+    void,
+    [msg: MessagesFromChild, workerId: number]
+  >
+
+  beforeAll(async () => {
+    store.dispatch({ type: `DELETE_CACHE`, cacheIsCorrupt: true })
```
Comment (LekoArts) [inline packages/gatsby/src/utils/worker/__tests__/jobs.ts:27]: We didn't use `cacheIsCorrupt` in the other tests, why here? :)

File: packages/gatsby/src/utils/worker/__tests__/jobs.ts [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+    const siteDirectory = path.join(__dirname, `fixtures`, `sample-site`)
+
+    await worker.all.loadConfigAndPlugins({
+      siteDirectory,
+      processFlags: false,
+    })
+
+    // plugin API creates a job
+    await Promise.all(worker.all.runAPI(`createSchemaCustomization`))
```
Comment (LekoArts) [inline packages/gatsby/src/utils/worker/__tests__/jobs.ts:43]: Mixing between using `Promise.all` and not?

File: packages/gatsby/src/utils/worker/__tests__/jobs.ts (hunk 4 above)
Comment (pieh) [reply packages/gatsby/src/utils/worker/__tests__/jobs.ts:27]: `jobs` reducer is specific in that it only clears its state when `cacheIsCorrupt` is set - https://github.com/gatsbyjs/gatsby/blob/53e5991c75e568934f402ae94420031d5f9f9102/packages/gatsby/src/redux/reducers/jobsv2.ts#L20-L21

File: packages/gatsby/src/utils/worker/__tests__/jobs.ts (hunk 5 above)
Comment (pieh) [reply packages/gatsby/src/utils/worker/__tests__/jobs.ts]: made it consistent (and correct as await previously wasn't doing anything) in https://github.com/gatsbyjs/gatsby/pull/32162/commits/73a31872d379b1d8e9fd111e32a7b340b26a6ea5

**Review by LekoArts — APPROVED** [review]

**Pull request merged**